driver-core/src/main/com/mongodb/internal/session/ServerSessionPool.java

```diff
@@ -94,9 +94,17 @@ public void close() {
         try {
             closing = true;
             serverSessionPool.close();
-            endClosedSessions();
+
+            List<BsonDocument> identifiers;
+            synchronized (this) {
+                identifiers = new ArrayList<BsonDocument>(closedSessionIdentifiers);
+            }
+            endClosedSessions(identifiers);
         } finally {
             closed = true;
+            synchronized (this) {
```

Comment:
This should moved out of the finally block and into the synchronized block within the try block.   Otherwise we may end up clearing elements that were added to the shared array between the copy and the here.